server/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java

```diff
@@ -351,13 +365,18 @@ static boolean executeBulkItemRequest(
             );
         } else {
             final IndexRequest request = context.getRequestToExecute();
+            DocumentParsingObserver documentParsingObserver = request.getNormalisedBytesParsed() != -1L
+                ? documentParsingSupplier.forAlreadyParsedInIngest(request.getNormalisedBytesParsed())
+                : documentParsingSupplier.getNewObserver();
```

Comment:
I could simplify this with just 
```suggestion
            DocumentParsingObserver documentParsingObserver = request.getNormalisedBytesParsed() != -1L
                ? DocumentParsingObserver.EMPTY_INSTANCE
                : documentParsingSupplier.getNewObserver();
```
Then in [onComplete](https://github.com/elastic/elasticsearch/pull/104859/files#diff-ba17a469a9179e83a5ff4a1fd6800866cd006de5235cf36523b008f6bf62909eR478 )
I could simply add an if
```
   DocumentParsingReporter documentParsingReporter = documentParsingSupplier.getDocumentParsingReporter();
            IndexRequest request = context.getRequestToExecute();
            if(request.getNormalisedBytesParsed()!=-1) {
                documentParsingReporter.onCompleted(docWriteRequest.index(), request.getNormalisedBytesParsed());
            } else {
                DocumentParsingObserver documentParsingObserver = context.getDocumentParsingObserver();
                documentParsingReporter.onCompleted(docWriteRequest.index(), documentParsingObserver.normalisedBytesParsed());
            }
```


with this I could remove the `forAlreadyParsedInIngest` method and the FixedDocumentParsingObserver instance

Reply:
Whether simplified or not, maybe it would be nice to extract this out into a static method on `DocumentParsingObserver` (or something like that)? It would give you a handy place to write a javadoc and would encapsulate the `-1L` magic number.

Reply:
fair point, better to hide the complexity of dealing with a magic number in a serverless plugin